### ISSUE TITLE
Link previews should respect robots.txt

### ISSUE THREAD

**Issue by jeffkaufman** [mastodon/mastodon#21738]
### Steps to reproduce the problem

1. Share https://www.jefftk.com/test/no-robots or another link which prohibits crawlers via robots.txt

### Expected behaviour

Mastodon instances fetch https://www.jefftk.com/robots.txt and don't fetch https://www.jefftk.com/test/no-robots

### Actual behaviour

Mastodon instances don't fetch https://www.jefftk.com/robots.txt and do fetch https://www.jefftk.com/test/no-robots

### Detailed description

An instance that is fetching pages because it heard about them via federation is operating as an automated agent, and so should respect robots.txt (https://www.rfc-editor.org/rfc/rfc9309.html).

Context: https://github.com/mastodon/mastodon/issues/4486, https://www.jefftk.com/p/mastodons-dubious-crawler-exemption, https://mastodon.mit.edu/@jefftk/109416209502343043

### Specifications

I see this in v4.0.2, but it's been around as long as Mastodon has had link previews

**Comment by nightpool** [comment]
Server authors do not write robots.txt files with social link previews in mind, because no other software has used robots.txt to govern its crawling before. Can you cite a practical use-case in the wild where a file that was blocked by robots.txt got crawled, and it was clear from author intent that it shouldn't have been and that link previews should be disabled for that route?

I don't think this is a practical stop-gap and I'm not sure what purpose it would serve for most users. And frankly, anyone sophisticated enough to think about editing their robots.txt files manually to block Mastodon link previews is probably already going to have a caching proxy in front of their content anyway. So I'm not sure who this would serve.

**Comment by nemobis** [comment]
Considering Mastodon only makes one request, this would double the number of requests made to servers. Is that really an improvement?

I notice that https://www.rfc-editor.org/rfc/rfc9309.html (draft) doesn't define *crawler*, but the one example it mentions is "crawlers to recursively traverse links for indexing".

**Comment by jeffkaufman** [comment]
> Can you cite a practical use-case in the wild where a file that was blocked by robots.txt got crawled, and it was clear from author intent that it shouldn't have been and that link previews should be disabled for that route?

This sounds backwards to me?  Since Mastodon doesn't currently respect robots.txt it's not surprising that people aren't trying to use robots.txt to restrict its behavior.

> anyone sophisticated enough to think about editing their robots.txt files manually to block Mastodon link previews is probably already going to have a caching proxy in front of their content anyway.

How about https://www.jwz.org/blog/2022/11/mastodon-stampede/ ?  And they do have social-media bot specific directives in https://www.jwz.org/robots.txt

> this would double the number of requests made to servers. Is that really an improvement?

Requests for robots.txt are generally far lighter to handle than ones for HTML.  The former is almost always a simple static file, while the latter often requires executing scripts and looking up things in databases.  Additionally, turning on caching for HTML can be tricky if your site has anything personalized while caching for robots.txt is totally safe.

Sticking with the example above, https://www.jwz.org/robots.txt fetches in about 100ms (mostly network delays), while https://www.jwz.org/blog/2022/11/mastodon-stampede/ and other posts take about a second (very likely db lookups).

**Comment by allo-** [comment]
Other crawlers use a cache for robots.txt. This wouldn't help with the initial request, but further robots.txt requests would be delayed by, e.g., a week.

But it seems to me to be an XY problem. One would expect a preview to be generated as a user would request a page and not as a crawler would, even showing previews for deep links that are not allowed for search engine crawlers.
The problem that needs to be solved is deduplicating the preview generation effort. If you solve it with robots.txt, you may end up with many sites blocking all Mastodon requests, because currently Mastodon is not relevant enough to bring many visitors, but hitting them with many requests to generate previews from all instances that see the link.

So the options seem to be:

- Do nothing and sites may start to block on user-agent basis or even block IPs of instances that hit them several times
- Implement robots.txt and get a blanket ban for Mastodon preview requests as a simple solution
- Implement some solution that avoid duplicate fetches (worth a separate issue)

**Comment by jeffkaufman** [comment]
> Implement some solution that avoid duplicate fetches (worth a separate issue)

I think that issue is https://github.com/mastodon/mastodon/issues/4486?

**Comment by fariparedes** [comment]
This seems to me like an issue germane to decentralized infrastructure with far wider consequences than mere link previewing. It's a very easy problem for Twitter to solve because it is centralized; no matter how huge Twitter gets, it's only one website and will only ever hit a resource once (not including any Twitter users who independently browse the link after that). Obviously it's extremely wasteful for hundreds of websites to download a bunch of data they don't need because it *might* be seen by a user and you want to serve it immediately without delay. Imagine if every website on the Internet were built like that (and after all, for many people, a totally decentralized Internet is indeed their goal); it seems like it would be an extremely unmanageable, unpleasant, and expensive affair for everyone involved.

I think that automated services respecting robots.txt is ideal and this suggestion is likely worth implementing. I don't think nemobis's concern really holds water, because requests are not created equally. robots.txt is, as has been previously mentioned, a lightweight static file easily served hundreds of times a second by any webserver worthy of being connected to the WWW, and it seems to me that while requests/sec have been bandied about a bit in discussions on this subject, the real problem is when [Mastodon hits something expensive](https://github.com/mastodon/mastodon/issues/4486#issuecomment-846626205); that's when things really break and people get upset.

Changing a HEAD+GET request to a GET request is nice but the amount of trouble it actually saves anyone is probably minuscule. Normally, Mastodon can't know anything about a random URL it's accessing until it's accessed it at least once; however, we can always know for certain that (under any normal operating circumstances) robots.txt is a lightweight text file, and it shouldn't do a server any harm to download it.

So I support this proposal on its face; *however* as a solution to the problem of decentralized infrastructure duplicating resources I think it is a band-aid on a gushing wound for the reasons laid out above and fixes nothing. Mastodon should probably respect robots.txt, but it hardly even qualifies as "doing something" about the problem of stampeding, if that's why this issue was opened. Anyone equipped to even realize this is a problem is more than capable of resolving it by blocking the user-agent as jwz did. Why jwz doesn't have a caching proxy is beyond me, maybe they have some principled opposition to Cloudflare, it doesn't really matter, because anyone down to the smallest Pi owner can block user-agents independently. Respecting robots.txt merely makes this solution slightly easier.

This is all I suppose a long-winded way of saying that I think contributors need to shift their paradigm from thinking about "how do we fix link previewing?" to "how do we deduplicate effort in our federated network as often as possible?" I would post it in 4486 as well but I don't want to resurrect a year+ old conversation that nobody's talking in. #21461 seems like a more promising solution albeit it comes with its own caveats since it attacks the problem directly.

**Comment by nemobis** [comment]
> Requests for robots.txt are generally far lighter to handle than ones for HTML.

That's true but assumes that a txt file is served. Many webservers will just return an HTML error page, which occasionally may be heavier than the target page. (This happened even to Wikipedia once!)

Also, the number of requests was enough of a concern that previously:

> I've already changed Mastodon's code to perform one GET instead of HEAD+GET, which should half the amount of requests any link gets from the fediverse with 3.0.

https://github.com/mastodon/mastodon/issues/4486#issuecomment-540152517

**Comment by smiba** [comment]
I would agree this is a worthy addition, blocking mastodon through robots.txt is a bit of an annoying move as it damages your link previews, but I don't think it's right we don't give administrators an easy way out of being hit by instances as soon as someone posts a link.
